vispy/scene/canvas.py

```diff
@@ -449,13 +461,20 @@ def visuals_at(self, pos, radius=10):
         visuals = [VisualNode._visual_ids.get(x, None) for x in ids]
         return [v for v in visuals if v is not None]
 
-    def _render_picking(self, **kwargs):
+    def _render_picking(self, crop):
```

Comment:
Why remove all other keyword arguments as an option? Also how does the overall resulting array change between using `region` versus `crop`?

Reply:
First off thanks for reviewing - I intentionally changed to a WIP to get your guys help in this as it isn't an easy undertaking. 

The background for the change is that the picking originally renders the ids to a fbo that is at the point clicked (some offset, 1x1 size) - this triggers a transform change for all visuals for every single mouse click (assuming the each click is at a different position). To retain the functionality of the public render() method I wanted region to work as before, so I need a way to crop out an area of the rendered frame buffer. I added the crop option to fbo reads, and since _render_picking is a private method I changed the way it works - now it requires specifying a crop to read out a set of pixels from a full fbo render (over the whole canvas). Thus crop specifies where out of the full set of ids I want a result, instead of the use of region keyword defining the area to render to (thus the kwargs are removed as the default of render is to deal with the whole canvas). One of the performance hits I was seeing in interacting with the examples/benchmark/scene_test_2.py program is the time spent doing transform updates for each click, even if nothing changed.

Given how long I spent in the nitty gritty of this I now feel I may of unnecessarily committed the change you are worried about in the transform_system.py below. The change vs. no change of the transform system during picking was born out of the fbo tranform changing often - but since the scene/canvas.py changes I made handle this (by fixing the picking rendering to the full canvas) I think that commit might be double useless (as you point out the STTransform already checks for changes).

Reply:
Correct me if I'm wrong: This change draws the entire canvas then only returns what is requested in the `crop` parameter. What would be the difference of only drawing the `region` needed and returning that whole region? I'm not sure you would need it but you could even pass `crop=region` although I don't think you would need to because it should be the correct size already.

I'm not sure I understood why the `**kwargs` had to go away. Even if you wanted `self, crop, **kwargs` you could still do that (although I think `self, **kwargs` is just as fine).

Reply:
Yes - this change has the whole canvas drawn and only the area specified by crop is read back for the picking ids.

The difference is that the original picking code would cause transform changes in the scene for every single click at a different location (i.e. the offset of the fbo). Specifically the calls to push_fbo()/pop_fbo() were running transform updates for every click. I found this much more expensive then just doing the render call to a full fbo and reading back the desired portion via crop (which avoids transform changes).

I'm not sure I have a good argument for the **kwargs change. It effectively removes flexibility to the _render_picking() call - that is _render_picking always renders to the full canvas and only allows the region of interest that is to be returned to be specified (i.e. crop). I can revert it back to **kwargs and pass in crop=(...), this will then allow full control over the call to render(), but any change to the region/size will cause transform updates to get triggered as before. I think the main reason I did it was to make sure that the render call always rendered to the full fbo. If I revert to the **kwargs I can leave a comment around the call to render() in _render_picking explaining the performance impact of modifying the region/etc for other developers.

Reply:
Good enough on the `**kwargs`. Leave it the way you have it. Less chance of someone accidentally causing a bug in the future.

I'm still not sure I understand the `region` versus `crop` necessity (sorry, I think I'm missing something). I understand why your changes to the TransformSystem are necessary if `push/pop_fbo` are causing transform updates. I don't understand the difference between using `region` and `crop` when you think about what the final result is. For example if you remove all your new `crop` stuff and just specified `region` (as it was before) then it will only draw that region and the frame buffer *should* be the same array that you get with your new code. Right? `push/pop_fbo` is called in either case and a FrameBuffer is created *every* time in both cases. The only difference is that in your implementation the entire canvas is drawn which seems wasteful.

Reply:
Yes I get the same array in the end though the original region based approach triggers explicit transform updates per picking event, while in the crop based approach no transforms are triggered related to actual fbo offset changes.

This is part of the problem, and the other is handled in the transform_system below. In the case below even though the crop changes ensure no unnecessary transform changes are made, (i.e. map_from, map_to should be the same), they still trigger update events as the map_to, map_from pair generate a scale in _set_st that differs from the internal state of the transform (i.e. that z scale factor I think comes from the camera).

Thus in summary:
1) using region and rendering to an fbo at (xoff, yoff, 1, 1) forces an explicit transform update (map_from, map_to change).
2) using the crop and rendering to the full fbo ensures no explicit transform updates are triggered, but the same end result is read out.
3) regardless of using number two (map_from, map_to) don't change, comparisons in the STTranform and the whole transform system still trigger updates for all visuals (that scale factor differs in z from what is coming in)

Initially I though the full fbo render would be slower but from my tests the transform system updates are much more expensive then the fbo operations (actual timing I was looking at showed the push_fbo(), pop_fbo() going from expensive operations to negligible in terms of timing, and it was all due to the tranform updates).

Reply:
As to your last paragraph about assuming the full fbo render would be slower, that's what I thought too. You've convinced me that this shouldn't be a problem. I keep trying to come up with something but they all revolve around the transform being changed programmatically...which would require a transform update anyway. ;)